GH-46683: [C++][Python] Add utf8_zero_fill compute function for sign-aware zero padding

iabhi4 opened the pull request:
### What does this PR do?
Adds a new compute kernel `utf8_zero_fill` to Arrow Compute module. This kernel zero-pads UTF-8 strings while preserving leading signs.

### Why is it needed?
Python's `str.zfill()` behavior is useful and expected in many data cleaning scenarios. Arrow lacked a direct equivalent.

### Implementation details
- Introduces `Utf8ZeroFillTransform` with dedicated `ZeroFillOptions`
- Registers `utf8_zero_fill` as a compute kernel
- Adds Python bindings and a clean alias: `utf8_zfill = utf8_zero_fill`
- Includes sign-aware padding logic

### Are these changes tested?
- C++ unit tests: `TestStringKernels, Utf8ZeroFill`
- Python tests: `test_utf8_zfill()`
Ran both tests
* GitHub Issue: #46683

iabhi4 commented:
@pitrou Just wanted to give you a heads-up that the implementation and tests are ready for review.
Please let me know if I’ve missed anything or if there’s a `changelog` entry or any documentation I should update as part of this PR.

AlenkaF commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -1051,6 +1051,25 @@ def test_pad():
+def test_utf8_zfill():
+    arr = pa.array(['A', 'AB', 'ABC', None])
+    assert pc.utf8_zfill(
+        arr, options=pc.PadOptions(width=3, padding='0')
```
This could work with `pc.utf8_zfill(arr, width=3)` or `pc.utf8_zfill(arr, 3)` if I am not mistaken?

AlenkaF reviewed (commented):
Thanks for the PR @iabhi4 !

The CI failures look connected. [For example](https://github.com/apache/arrow/actions/runs/15659875844/job/44116061600?pr=46815#step:6:6058):

```
/arrow/cpp/src/arrow/compute/kernels/scalar_string_internal.h:179:  Check failed: _s.ok() Operation failed: registry->AddFunction(std::move(func))
Bad status: Invalid: In function 'utf8_zfill': description line length exceeds 78 characters
```

We do not require changelog entry per PR - it is done automatically at the release. As for the documentation, on the Python side, the User Guide is very minimal at this point and so the work would be out of scope of this PR (will keep in mind to maybe add a new issue under https://github.com/apache/arrow/issues/46601).

That being said, I would love if we would have a short example here: https://arrow.apache.org/docs/dev/python/generated/pyarrow.compute.PadOptions.html, even if only for the new `utf8_zfill` kernel 😊

AlenkaF commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -1051,6 +1051,25 @@ def test_pad():
     assert pc.utf8_rpad(arr, 3).tolist() == [None, 'á  ', 'abcd']
 
 
+def test_utf8_zfill():
```
Could we add a test where we do not use padding, only width, without the `pc.PadOptions`?

iabhi4 replied:
Just wanted to clarify this before I made the changes! Currently, the `utf8_zfill` kernel requires `PadOptions` (similar to how utf8_lpad behave), and there’s no overload that accepts just `width=` as a direct keyword argument.

Would you prefer that I add a test case that intentionally omits `PadOptions` to validate that a clear error is raised (like a `TypeError`)? Or were you referring to a different kind of test behavior? Thanks again for the guidance! 🙏

AlenkaF replied:
Thank you for the comment! My main intention was to understand whether adding a direct keyword is valid, as I noticed it's commonly used in tests for other string padding functions. That's clear now—thanks again! And yes, I'd be happy to see a test case that validates a `TypeError`.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_test.cc:
```diff
@@ -2241,6 +2241,32 @@ TYPED_TEST(TestStringKernels, PadUTF8) {
+  options.width = 2;
+  this->CheckUnary("utf8_zfill", R"(["AB", "-12", "+12", "XYZ"])", this->type(),
+                   R"(["AB", "-12", "+12", "XYZ"])", &options);
+
```
Can you please add tests with another padding character than "0"?

pitrou replied:
Also, can you add tests with non-ASCII characters (both in padding string and input strings)?

iabhi4 replied:
Added the additional tests as suggested

iabhi4 commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,13 +987,70 @@ const FunctionDoc utf8_rpad_doc(
+using Utf8ZFill = StringTransformExecWithState<Type, Utf8ZFillTransform>;
+
+const FunctionDoc utf8_zfill_doc(
+    "Left-pad strings with zeros, handling signs like Python str.zfill()",
```
Makes sense, removing the explicit mention of Python

iabhi4 commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,13 +987,70 @@ const FunctionDoc utf8_rpad_doc(
+using Utf8ZFill = StringTransformExecWithState<Type, Utf8ZFillTransform>;
+
+const FunctionDoc utf8_zfill_doc(
+    "Left-pad strings with zeros, handling signs like Python str.zfill()",
```
Yes the default is space

AlenkaF commented:
I think a suggestion of mine might have been lost during the review stage. What I meant to propose was setting `padding="0"` as the default, while still allowing users to override it with any other character. That would align better with the function's name (zero fill) and also provide the desired behaviour without requiring users to explicitly specify the padding each time.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,13 +987,77 @@ const FunctionDoc utf8_rpad_doc(
+    }
+    int64_t num_zeros = spaces;
+    std::string padding = options_.padding;
+    if (padding == " ") {
+      padding = "0";
+    }
```
As @AlenkaF pointed out, we can't do that because it prevents the user from having a whitespace-filled zero padding.

If we really want "0" as a default, we should therefore use a separate `ZeroFillOptions`.

pitrou commented:
Time for naming discussion: should the discussion be called "utf8_zfill" (Python inspiration, but cryptic to non-Python programmers) or "utf8_zero_fill" (longer, more explicit)? cc @zanmato1984

zanmato1984 commented:
Is it possible to use name `utf8_zero_fill` in C++ while map it to `utf8_zfill` in pyarrow?

pitrou commented:
We could, but for now we're mirroring the function names exactly, so that could end up confusing. Also, the documentation would be less shareable between C++ and Python.

pitrou commented:
We could add a Python alias though, such as `utf8_zfill = utf8_zero_fill`.

zanmato1984 commented:
Aliasing sounds ideal I guess?

zanmato1984 commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,13 +987,77 @@ const FunctionDoc utf8_rpad_doc(
+    }
+    return Status::OK();
+  }
+  int64_t MaxCodeunits(int64_t ninputs, int64_t input_ncodeunits) override {
```
```suggestion

  int64_t MaxCodeunits(int64_t ninputs, int64_t input_ncodeunits) override {
```

zanmato1984 commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,13 +987,77 @@ const FunctionDoc utf8_rpad_doc(
                                                   utf8_center_doc);
+  MakeUnaryStringBatchKernelWithState<Utf8ZFill>("utf8_zfill", registry, utf8_zfill_doc);
 }
-
```
Why removing this blank line?

iabhi4 commented:
Just pushed the final changes. I introduced `ZeroFillOptions` as a standalone class instead of overloading `PadOptions`, renamed the kernel to `utf8_zero_fill` for clarity, and added a clean alias in Python `(utf8_zfill = utf8_zero_fill)`

Tests and docs are updated accordingly. Let me know if there's anything else you'd like adjusted

pitrou commented on cpp/src/arrow/compute/kernels/scalar_string_utf8.cc:
```diff
@@ -987,11 +987,74 @@ const FunctionDoc utf8_rpad_doc(
+      input++;
+      input_string_ncodeunits--;
+    }
+    int64_t num_zeros = spaces;
```
Why not call `spaces` `num_zeros` from the start?

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -1142,6 +1142,36 @@ class PadOptions(_PadOptions):
+    width : int
+        Desired string length.
+    padding : str, default "0"
+        Padding character. Should be one byte or Unicode codepoint.
```
There is no Ascii version of this function, so we can simplify the description just like in C++.
```suggestion
        Padding character. Should be one Unicode codepoint.
```

pitrou commented on python/pyarrow/compute.py:
```diff
@@ -338,6 +339,8 @@ def _make_global_functions():
 
 _make_global_functions()
+# Alias for consistency
+utf8_zfill = utf8_zero_fill = globals()["utf8_zero_fill"]
```
This can be simplified to:
```suggestion
utf8_zfill = utf8_zero_fill
```

pitrou replied:
Ah, I see. Can you add a comment that it's required for linting, then?

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -1051,6 +1052,31 @@ def test_pad():
 
+def test_utf8_zfill():
+    assert pc.utf8_zfill is pc.utf8_zero_fill
+    arr = pa.array(['A', 'AB', 'ABC', None])
```
Instead of writing the test examples by hand, perhaps validate that `pc.utf8_zero_fill` has the same semantics as `str.zfill`?

pitrou reviewed (changes requested):
Thanks a lot for the update @iabhi4 . Here is a small number of additional suggestions.

iabhi4 commented:
Thanks for the feedback @pitrou, I’ve addressed all the suggestions!

pitrou commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -167,6 +167,7 @@ def test_option_class_equality(request):
         pc.ModeOptions(),
         pc.NullOptions(),
         pc.PadOptions(5),
+        pc.ZeroFillOptions(4, "0"),
```
Same here: keep it alphabetically-ordered?

kou commented:
@github-actions crossbow submit test-ubuntu-22.04-cpp-20
